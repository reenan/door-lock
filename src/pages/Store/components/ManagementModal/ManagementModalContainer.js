import React, { Component } from 'react'
import PropTypes from 'prop-types'
import uuid4 from 'uuid4'

import ManagementModal from './ManagementModal'

class ManagementModalContainer extends Component {
  constructor(props) {
    super(props)

    this.state = {
      // Create a copy from list removing reference to use inside modal
      virtualItems: JSON.parse(JSON.stringify(this.props.items))
    }
  }

  // Generically adds a new item to the list
  addNewItem = () => {
    const { virtualItems } = this.state
    const newUUID = uuid4()

    // Add new item based on passed structure
    const { itemStructure } = this.props
    virtualItems[newUUID] = Object.assign({}, itemStructure)

    this.setState({ virtualItems })
  }

  // Generic handles text input modification on item attribute
  handleTextInputChange = (id, event) => {
    const { virtualItems } = this.state
    const { name, value } = event.target

    virtualItems[id][name] = value

    this.setState({ virtualItems })
  }

  // Generic handles checkbox toggle on item attribute
  // Logic "hard-coded" for checkboxes regarding nested data
  handleCheckboxToggle = (id, nestedID, attribute) => {
    const { virtualItems } = this.state

    virtualItems[id][attribute][nestedID] = !virtualItems[id][attribute][nestedID]

    this.setState({ virtualItems })
  }

  // Logic hard-coded to change element ROLE attr
  // TODO: remove hard-coded ROLE attr
  handleDropdownChange = (id, _event, { value }) => {
    let { virtualItems } = this.state
    virtualItems[id].role = value

    this.setState({ virtualItems })
  }

  // Generically removes an item from the list
  removeItem = (id) => {
    const { virtualItems } = this.state
    delete virtualItems[id]

    this.setState({ virtualItems })
  }

  // Discard changes made into the modal
  discardChanges = () => {

    this.setState({
      virtualItems: JSON.parse(JSON.stringify(this.props.items))
    }, this.props.close)

  }

  // Save modal changes on data
  saveChanges = () => {
    const { virtualItems } = this.state

    // Remove items that does not have a name
    const cleanedItemList = Object.keys(virtualItems).reduce((cleanList, id) => {
      if (virtualItems[id].name !== '') {
        cleanList[id] = virtualItems[id]
      }

      return cleanList
    }, {})

    // Update state then propagate changes
    this.setState({ virtualItems: cleanedItemList }, () => {

      // Remove reference to list before passing list to redux
      this.props.save(JSON.parse(JSON.stringify(cleanedItemList))).then(() => {
        this.props.close()
      })
    })
  }

  render() {
    const {
      isOpen,
      close,
      loading,
      modalData,
      useDivider,
      nestedItems,
      ItemComponent
    } = this.props

    const { virtualItems } = this.state
    const virtualItemsKeys = Object.keys(virtualItems)
    const nestedItemsKeys = Object.keys(nestedItems)

    return (
      <ManagementModal
        isOpen={isOpen}
        close={close}
        loading={loading}
        useDivider={useDivider}
        modalData={modalData}
        nestedItems={nestedItems}
        nestedItemsKeys={nestedItemsKeys}
        virtualItems={virtualItems}
        virtualItemsKeys={virtualItemsKeys}
        ItemComponent={ItemComponent}
        handleDropdownChange={this.handleDropdownChange}
        handleCheckboxToggle={this.handleCheckboxToggle}
        handleTextInputChange={this.handleTextInputChange}
        removeItem={this.removeItem}
        addNewItem={this.addNewItem}
        discardChanges={this.discardChanges}
        saveChanges={this.saveChanges}
      />
    )
  }
}

ManagementModalContainer.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
  save: PropTypes.func.isRequired,
  items: PropTypes.object.isRequired,
  nestedItems: PropTypes.object,
  itemStructure: PropTypes.object.isRequired,
  modalData: PropTypes.object.isRequired,
  loading: PropTypes.bool,
  useDivider: PropTypes.bool,
}

ManagementModalContainer.defaultProps = {
  loading: false,
  useDivider: false,
  nestedItems: {},
}

export default ManagementModalContainer
