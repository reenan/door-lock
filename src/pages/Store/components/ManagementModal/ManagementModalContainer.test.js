import React from 'react'
import ManagementModalContainer from './ManagementModalContainer'
import { shallow } from 'enzyme'

import EnzymeToJson from 'enzyme-to-json'
import { mount } from 'enzyme'

const MockComponent = () => (
  <div>
    Hello Test
  </div>
)

it('renders without crashing', () => {
  const component = mount(
    <ManagementModalContainer isOpen={false} items={{}} nestedItems={{}}
      loading={false} save={() => {}} close={() => {}} itemStructure={{ name: '' }}
      useDivider ItemComponent={MockComponent}
      modalData={{
        title: `Mock`,
        description: `Write Tests`,
        info: `Its important.`,
        addText: 'Add new test',
      }}
    />
  )

  expect(EnzymeToJson(component)).toMatchSnapshot()
})

it('renders without crashing when open', () => {
  const component = mount(
    <ManagementModalContainer isOpen={true} items={{}} nestedItems={{}}
      loading={false} save={() => {}} close={() => {}} itemStructure={{ name: '' }}
      useDivider ItemComponent={MockComponent}
      modalData={{
        title: `Mock`,
        description: `Write Tests`,
        info: `Its important.`,
        addText: 'Add new test',
      }}
    />
  )

  expect(EnzymeToJson(component)).toMatchSnapshot()
})

it('should be able to add new item and remove it', () => {
  const component = shallow(
    <ManagementModalContainer isOpen={false} items={{}} nestedItems={{}}
      loading={false} save={() => {}} close={() => {}} itemStructure={{ name: '' }}
      useDivider ItemComponent={MockComponent}
      modalData={{
        title: `Mock`,
        description: `Write Tests`,
        info: `Its important.`,
        addText: 'Add new test',
      }}
    />
  )

  component.instance().addNewItem()

  const virtualItemsState = component.state('virtualItems')
  const virtualItemsKey = Object.keys(virtualItemsState)

  expect(virtualItemsKey.length).toBe(1)

  const itemID = virtualItemsKey[0]
  component.instance().removeItem(itemID)

  const newVirtualItemsState = component.state('virtualItems')
  const newVirtualItemsKey = Object.keys(newVirtualItemsState)

  expect(newVirtualItemsKey.length).toBe(0)
})

it('should be able to handle input changes', () => {
  const component = shallow(
    <ManagementModalContainer isOpen={false} items={{}} nestedItems={{}}
      loading={false} save={() => {}} close={() => {}}
      itemStructure={{
        name: '',
        nested: {
          toggle: false
        },
        role: null
      }}
      useDivider ItemComponent={MockComponent}
      modalData={{
        title: `Mock`,
        description: `Write Tests`,
        info: `Its important.`,
        addText: 'Add new test',
      }}
    />
  )

  component.instance().addNewItem()

  const virtualItemsState = component.state('virtualItems')
  const virtualItemsKey = Object.keys(virtualItemsState)

  const itemID = virtualItemsKey[0]

  component.instance().handleTextInputChange(itemID, {
    target: { name: 'name', value: 'mock value' }
  })
  component.instance().handleCheckboxToggle(itemID, 'toggle', 'nested')
  component.instance().handleDropdownChange(itemID, null, { value: 'mock-dropdown' })

  expect(virtualItemsState[itemID].name).toBe('mock value')
  expect(virtualItemsState[itemID].nested.toggle).toBe(true)
  expect(virtualItemsState[itemID].role).toBe('mock-dropdown')
})

it('should be able to discard changes and reset list', () => {
  const component = shallow(
    <ManagementModalContainer isOpen={false} items={{}} nestedItems={{}}
      loading={false} save={() => {}} close={() => {}}
      itemStructure={{ name: '' }}
      useDivider ItemComponent={MockComponent}
      modalData={{
        title: `Mock`,
        description: `Write Tests`,
        info: `Its important.`,
        addText: 'Add new test',
      }}
    />
  )

  component.instance().addNewItem()

  const virtualItemsState = component.state('virtualItems')
  const virtualItemsKey = Object.keys(virtualItemsState)

  const itemID = virtualItemsKey[0]
  expect(component.state('virtualItems')[itemID]).not.toBe(undefined)

  component.instance().discardChanges()
  expect(component.state('virtualItems')[itemID]).toBe(undefined)
})

it('should call props.save on saveChanges', () => {
  const saveMock = jest.fn()
  saveMock.mockReturnValue({
    then: (fn) => {
      fn()
    }
  })

  const component = shallow(
    <ManagementModalContainer isOpen={false}
      items={{
        mock: {
          name: 'mocked'
        }
      }}
      nestedItems={{}}
      loading={false} save={saveMock} close={() => {}}
      itemStructure={{ name: '' }}
      useDivider ItemComponent={MockComponent}
      modalData={{
        title: `Mock`,
        description: `Write Tests`,
        info: `Its important.`,
        addText: 'Add new test',
      }}
    />
  )

  component.instance().addNewItem()
  component.instance().saveChanges()

  expect(saveMock).toHaveBeenCalled()
})
