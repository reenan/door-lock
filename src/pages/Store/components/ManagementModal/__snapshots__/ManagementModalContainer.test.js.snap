// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<ManagementModalContainer
  ItemComponent={[Function]}
  close={[Function]}
  isOpen={false}
  itemStructure={
    Object {
      "name": "",
    }
  }
  items={Object {}}
  loading={false}
  modalData={
    Object {
      "addText": "Add new test",
      "description": "Write Tests",
      "info": "Its important.",
      "title": "Mock",
    }
  }
  nestedItems={Object {}}
  save={[Function]}
  useDivider={true}
>
  <_default
    ItemComponent={[Function]}
    addNewItem={[Function]}
    close={[Function]}
    discardChanges={[Function]}
    handleCheckboxToggle={[Function]}
    handleDropdownChange={[Function]}
    handleTextInputChange={[Function]}
    isOpen={false}
    loading={false}
    modalData={
      Object {
        "addText": "Add new test",
        "description": "Write Tests",
        "info": "Its important.",
        "title": "Mock",
      }
    }
    nestedItems={Object {}}
    nestedItemsKeys={Array []}
    removeItem={[Function]}
    saveChanges={[Function]}
    useDivider={true}
    virtualItems={Object {}}
    virtualItemsKeys={Array []}
  >
    <Modal
      centered={true}
      className="management-modal"
      closeOnDimmerClick={false}
      closeOnDocumentClick={false}
      closeOnEscape={false}
      dimmer={true}
      eventPool="Modal"
      onClose={[Function]}
      open={false}
      size="small"
    >
      <Portal
        closeOnDocumentClick={false}
        closeOnEscape={false}
        eventPool="Modal"
        mountNode={<body />}
        onClose={[Function]}
        onMount={[Function]}
        onOpen={[Function]}
        onUnmount={[Function]}
        open={false}
        openOnTriggerClick={true}
      />
    </Modal>
  </_default>
</ManagementModalContainer>
`;

exports[`renders without crashing when open 1`] = `
<ManagementModalContainer
  ItemComponent={[Function]}
  close={[Function]}
  isOpen={true}
  itemStructure={
    Object {
      "name": "",
    }
  }
  items={Object {}}
  loading={false}
  modalData={
    Object {
      "addText": "Add new test",
      "description": "Write Tests",
      "info": "Its important.",
      "title": "Mock",
    }
  }
  nestedItems={Object {}}
  save={[Function]}
  useDivider={true}
>
  <_default
    ItemComponent={[Function]}
    addNewItem={[Function]}
    close={[Function]}
    discardChanges={[Function]}
    handleCheckboxToggle={[Function]}
    handleDropdownChange={[Function]}
    handleTextInputChange={[Function]}
    isOpen={true}
    loading={false}
    modalData={
      Object {
        "addText": "Add new test",
        "description": "Write Tests",
        "info": "Its important.",
        "title": "Mock",
      }
    }
    nestedItems={Object {}}
    nestedItemsKeys={Array []}
    removeItem={[Function]}
    saveChanges={[Function]}
    useDivider={true}
    virtualItems={Object {}}
    virtualItemsKeys={Array []}
  >
    <Modal
      centered={true}
      className="management-modal"
      closeOnDimmerClick={false}
      closeOnDocumentClick={false}
      closeOnEscape={false}
      dimmer={true}
      eventPool="Modal"
      onClose={[Function]}
      open={true}
      size="small"
    >
      <Portal
        closeOnDocumentClick={false}
        closeOnEscape={false}
        eventPool="Modal"
        mountNode={
          <body
            class="dimmable dimmed"
          >
            <div
              class="ui page modals dimmer transition visible active"
              style="display: flex;"
            >
              <div
                class="ui small modal transition visible active management-modal"
              >
                <div
                  class="header"
                >
                  Mock
                </div>
                <div
                  class="content"
                >
                  <p>
                    Write Tests
                  </p>
                  <p>
                    Its important.
                  </p>
                  <div
                    class="ui middle aligned list item-list"
                    role="list"
                  >
                    <div
                      class="item fade visible transition"
                      role="listitem"
                    >
                      <div
                        class="right floated content"
                      >
                        <button
                          class="ui button"
                        >
                          <i
                            aria-hidden="true"
                            class="plus icon"
                          />
                          Add new test
                        </button>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ui dimmer"
                  >
                    <div
                      class="content"
                    >
                      <div
                        class="ui loader"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="actions"
                >
                  <button
                    class="ui button"
                  >
                    Discard
                  </button>
                  <button
                    class="ui positive button"
                  >
                    Save
                  </button>
                </div>
              </div>
            </div>
          </body>
        }
        onClose={[Function]}
        onMount={[Function]}
        onOpen={[Function]}
        onUnmount={[Function]}
        open={true}
        openOnTriggerClick={true}
      >
        <PortalInner
          innerRef={
            Object {
              "current": <div
                class="ui page modals dimmer transition visible active"
                style="display: flex;"
              >
                <div
                  class="ui small modal transition visible active management-modal"
                >
                  <div
                    class="header"
                  >
                    Mock
                  </div>
                  <div
                    class="content"
                  >
                    <p>
                      Write Tests
                    </p>
                    <p>
                      Its important.
                    </p>
                    <div
                      class="ui middle aligned list item-list"
                      role="list"
                    >
                      <div
                        class="item fade visible transition"
                        role="listitem"
                      >
                        <div
                          class="right floated content"
                        >
                          <button
                            class="ui button"
                          >
                            <i
                              aria-hidden="true"
                              class="plus icon"
                            />
                            Add new test
                          </button>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ui dimmer"
                    >
                      <div
                        class="content"
                      >
                        <div
                          class="ui loader"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="actions"
                  >
                    <button
                      class="ui button"
                    >
                      Discard
                    </button>
                    <button
                      class="ui positive button"
                    >
                      Save
                    </button>
                  </div>
                </div>
              </div>,
            }
          }
          mountNode={
            <body
              class="dimmable dimmed"
            >
              <div
                class="ui page modals dimmer transition visible active"
                style="display: flex;"
              >
                <div
                  class="ui small modal transition visible active management-modal"
                >
                  <div
                    class="header"
                  >
                    Mock
                  </div>
                  <div
                    class="content"
                  >
                    <p>
                      Write Tests
                    </p>
                    <p>
                      Its important.
                    </p>
                    <div
                      class="ui middle aligned list item-list"
                      role="list"
                    >
                      <div
                        class="item fade visible transition"
                        role="listitem"
                      >
                        <div
                          class="right floated content"
                        >
                          <button
                            class="ui button"
                          >
                            <i
                              aria-hidden="true"
                              class="plus icon"
                            />
                            Add new test
                          </button>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ui dimmer"
                    >
                      <div
                        class="content"
                      >
                        <div
                          class="ui loader"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="actions"
                  >
                    <button
                      class="ui button"
                    >
                      Discard
                    </button>
                    <button
                      class="ui positive button"
                    >
                      Save
                    </button>
                  </div>
                </div>
              </div>
            </body>
          }
          onMount={[Function]}
          onUnmount={[Function]}
        >
          <Portal
            containerInfo={
              <body
                class="dimmable dimmed"
              >
                <div
                  class="ui page modals dimmer transition visible active"
                  style="display: flex;"
                >
                  <div
                    class="ui small modal transition visible active management-modal"
                  >
                    <div
                      class="header"
                    >
                      Mock
                    </div>
                    <div
                      class="content"
                    >
                      <p>
                        Write Tests
                      </p>
                      <p>
                        Its important.
                      </p>
                      <div
                        class="ui middle aligned list item-list"
                        role="list"
                      >
                        <div
                          class="item fade visible transition"
                          role="listitem"
                        >
                          <div
                            class="right floated content"
                          >
                            <button
                              class="ui button"
                            >
                              <i
                                aria-hidden="true"
                                class="plus icon"
                              />
                              Add new test
                            </button>
                          </div>
                        </div>
                      </div>
                      <div
                        class="ui dimmer"
                      >
                        <div
                          class="content"
                        >
                          <div
                            class="ui loader"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="actions"
                    >
                      <button
                        class="ui button"
                      >
                        Discard
                      </button>
                      <button
                        class="ui positive button"
                      >
                        Save
                      </button>
                    </div>
                  </div>
                </div>
              </body>
            }
          >
            <Ref
              innerRef={[Function]}
            >
              <RefFindNode
                innerRef={[Function]}
              >
                <div
                  className="ui page modals dimmer transition visible active"
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <div
                          class="ui small modal transition visible active management-modal"
                        >
                          <div
                            class="header"
                          >
                            Mock
                          </div>
                          <div
                            class="content"
                          >
                            <p>
                              Write Tests
                            </p>
                            <p>
                              Its important.
                            </p>
                            <div
                              class="ui middle aligned list item-list"
                              role="list"
                            >
                              <div
                                class="item fade visible transition"
                                role="listitem"
                              >
                                <div
                                  class="right floated content"
                                >
                                  <button
                                    class="ui button"
                                  >
                                    <i
                                      aria-hidden="true"
                                      class="plus icon"
                                    />
                                    Add new test
                                  </button>
                                </div>
                              </div>
                            </div>
                            <div
                              class="ui dimmer"
                            >
                              <div
                                class="content"
                              >
                                <div
                                  class="ui loader"
                                />
                              </div>
                            </div>
                          </div>
                          <div
                            class="actions"
                          >
                            <button
                              class="ui button"
                            >
                              Discard
                            </button>
                            <button
                              class="ui positive button"
                            >
                              Save
                            </button>
                          </div>
                        </div>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <div
                            class="ui small modal transition visible active management-modal"
                          >
                            <div
                              class="header"
                            >
                              Mock
                            </div>
                            <div
                              class="content"
                            >
                              <p>
                                Write Tests
                              </p>
                              <p>
                                Its important.
                              </p>
                              <div
                                class="ui middle aligned list item-list"
                                role="list"
                              >
                                <div
                                  class="item fade visible transition"
                                  role="listitem"
                                >
                                  <div
                                    class="right floated content"
                                  >
                                    <button
                                      class="ui button"
                                    >
                                      <i
                                        aria-hidden="true"
                                        class="plus icon"
                                      />
                                      Add new test
                                    </button>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="ui dimmer"
                              >
                                <div
                                  class="content"
                                >
                                  <div
                                    class="ui loader"
                                  />
                                </div>
                              </div>
                            </div>
                            <div
                              class="actions"
                            >
                              <button
                                class="ui button"
                              >
                                Discard
                              </button>
                              <button
                                class="ui positive button"
                              >
                                Save
                              </button>
                            </div>
                          </div>,
                        }
                      }
                    >
                      <div
                        className="ui small modal transition visible active management-modal"
                        style={
                          Object {
                            "marginTop": null,
                          }
                        }
                      >
                        <MountNode
                          className="dimmable dimmed"
                        />
                        <ModalHeader>
                          <div
                            className="header"
                          >
                            Mock
                          </div>
                        </ModalHeader>
                        <ModalContent>
                          <div
                            className="content"
                          >
                            <p>
                              Write Tests
                            </p>
                            <p>
                              Its important.
                            </p>
                            <TransitionGroup
                              animation="fade"
                              as={[Function]}
                              className="item-list"
                              duration={200}
                              verticalAlign="middle"
                            >
                              <List
                                className="item-list"
                                verticalAlign="middle"
                              >
                                <div
                                  className="ui middle aligned list item-list"
                                  role="list"
                                >
                                  <Transition
                                    animation="fade"
                                    duration={200}
                                    key=".1"
                                    mountOnShow={true}
                                    onHide={[Function]}
                                    reactKey=".1"
                                    transitionOnMount={false}
                                    unmountOnHide={false}
                                    visible={true}
                                  >
                                    <ListItem
                                      className="fade visible transition"
                                      key=".1"
                                      style={
                                        Object {
                                          "animationDuration": undefined,
                                        }
                                      }
                                    >
                                      <div
                                        className="item fade visible transition"
                                        onClick={[Function]}
                                        role="listitem"
                                        style={
                                          Object {
                                            "animationDuration": undefined,
                                          }
                                        }
                                      >
                                        <ListContent
                                          floated="right"
                                        >
                                          <div
                                            className="right floated content"
                                          >
                                            <Button
                                              as="button"
                                              content="Add new test"
                                              icon="plus"
                                              onClick={[Function]}
                                            >
                                              <Ref
                                                innerRef={
                                                  Object {
                                                    "current": <button
                                                      class="ui button"
                                                    >
                                                      <i
                                                        aria-hidden="true"
                                                        class="plus icon"
                                                      />
                                                      Add new test
                                                    </button>,
                                                  }
                                                }
                                              >
                                                <RefFindNode
                                                  innerRef={
                                                    Object {
                                                      "current": <button
                                                        class="ui button"
                                                      >
                                                        <i
                                                          aria-hidden="true"
                                                          class="plus icon"
                                                        />
                                                        Add new test
                                                      </button>,
                                                    }
                                                  }
                                                >
                                                  <button
                                                    className="ui button"
                                                    onClick={[Function]}
                                                  >
                                                    <Icon
                                                      as="i"
                                                      name="plus"
                                                    >
                                                      <i
                                                        aria-hidden="true"
                                                        className="plus icon"
                                                        onClick={[Function]}
                                                      />
                                                    </Icon>
                                                    Add new test
                                                  </button>
                                                </RefFindNode>
                                              </Ref>
                                            </Button>
                                          </div>
                                        </ListContent>
                                      </div>
                                    </ListItem>
                                  </Transition>
                                </div>
                              </List>
                            </TransitionGroup>
                            <Dimmer
                              active={false}
                            >
                              <DimmerInner
                                active={false}
                              >
                                <Ref
                                  innerRef={
                                    Object {
                                      "current": <div
                                        class="ui dimmer"
                                      >
                                        <div
                                          class="content"
                                        >
                                          <div
                                            class="ui loader"
                                          />
                                        </div>
                                      </div>,
                                    }
                                  }
                                >
                                  <RefFindNode
                                    innerRef={
                                      Object {
                                        "current": <div
                                          class="ui dimmer"
                                        >
                                          <div
                                            class="content"
                                          >
                                            <div
                                              class="ui loader"
                                            />
                                          </div>
                                        </div>,
                                      }
                                    }
                                  >
                                    <div
                                      className="ui dimmer"
                                      onClick={[Function]}
                                    >
                                      <div
                                        className="content"
                                      >
                                        <Loader>
                                          <div
                                            className="ui loader"
                                          />
                                        </Loader>
                                      </div>
                                    </div>
                                  </RefFindNode>
                                </Ref>
                              </DimmerInner>
                            </Dimmer>
                          </div>
                        </ModalContent>
                        <ModalActions>
                          <div
                            className="actions"
                          >
                            <Button
                              as="button"
                              content="Discard"
                              onClick={[Function]}
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui button"
                                    >
                                      Discard
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui button"
                                      >
                                        Discard
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui button"
                                    onClick={[Function]}
                                  >
                                    Discard
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                            <Button
                              as="button"
                              content="Save"
                              onClick={[Function]}
                              positive={true}
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui positive button"
                                    >
                                      Save
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui positive button"
                                      >
                                        Save
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui positive button"
                                    onClick={[Function]}
                                  >
                                    Save
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                          </div>
                        </ModalActions>
                      </div>
                    </RefFindNode>
                  </Ref>
                </div>
              </RefFindNode>
            </Ref>
          </Portal>
        </PortalInner>
        <EventStack
          name="mouseleave"
          on={[Function]}
          pool="Modal"
          target={
            Object {
              "current": <div
                class="ui page modals dimmer transition visible active"
                style="display: flex;"
              >
                <div
                  class="ui small modal transition visible active management-modal"
                >
                  <div
                    class="header"
                  >
                    Mock
                  </div>
                  <div
                    class="content"
                  >
                    <p>
                      Write Tests
                    </p>
                    <p>
                      Its important.
                    </p>
                    <div
                      class="ui middle aligned list item-list"
                      role="list"
                    >
                      <div
                        class="item fade visible transition"
                        role="listitem"
                      >
                        <div
                          class="right floated content"
                        >
                          <button
                            class="ui button"
                          >
                            <i
                              aria-hidden="true"
                              class="plus icon"
                            />
                            Add new test
                          </button>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ui dimmer"
                    >
                      <div
                        class="content"
                      >
                        <div
                          class="ui loader"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="actions"
                  >
                    <button
                      class="ui button"
                    >
                      Discard
                    </button>
                    <button
                      class="ui positive button"
                    >
                      Save
                    </button>
                  </div>
                </div>
              </div>,
            }
          }
        />
        <EventStack
          name="mouseenter"
          on={[Function]}
          pool="Modal"
          target={
            Object {
              "current": <div
                class="ui page modals dimmer transition visible active"
                style="display: flex;"
              >
                <div
                  class="ui small modal transition visible active management-modal"
                >
                  <div
                    class="header"
                  >
                    Mock
                  </div>
                  <div
                    class="content"
                  >
                    <p>
                      Write Tests
                    </p>
                    <p>
                      Its important.
                    </p>
                    <div
                      class="ui middle aligned list item-list"
                      role="list"
                    >
                      <div
                        class="item fade visible transition"
                        role="listitem"
                      >
                        <div
                          class="right floated content"
                        >
                          <button
                            class="ui button"
                          >
                            <i
                              aria-hidden="true"
                              class="plus icon"
                            />
                            Add new test
                          </button>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ui dimmer"
                    >
                      <div
                        class="content"
                      >
                        <div
                          class="ui loader"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="actions"
                  >
                    <button
                      class="ui button"
                    >
                      Discard
                    </button>
                    <button
                      class="ui positive button"
                    >
                      Save
                    </button>
                  </div>
                </div>
              </div>,
            }
          }
        />
        <EventStack
          name="mousedown"
          on={[Function]}
          pool="Modal"
          target="document"
        />
        <EventStack
          name="click"
          on={[Function]}
          pool="Modal"
          target="document"
        />
        <EventStack
          name="keydown"
          on={[Function]}
          pool="Modal"
          target="document"
        />
      </Portal>
    </Modal>
  </_default>
</ManagementModalContainer>
`;
